File: lib/domain/cultivation_plan/mappers/plan_save_fertilize_unique_name.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Mappers
      class PlanSaveFertilizeUniqueName
        # @yieldparam candidate [String]
        def self.each_candidate(base_name, &block)
          block.call("#{base_name} (コピー)")
          suffix = 2
          loop do
            block.call("#{base_name} (コピー #{suffix})")
            suffix += 1
          end
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/mappers/plan_save_fertilize_unique_name.rs
//! Ruby: `Domain::CultivationPlan::Mappers::PlanSaveFertilizeUniqueName`

pub fn resolve_unique_name<F>(base_name: &str, mut exists: F) -> Option<String>
where
    F: FnMut(&str) -> bool,
{
    let first = format!("{base_name} (コピー)");
    if !exists(&first) {
        return Some(first);
    }
    let mut suffix = 2;
    loop {
        let candidate = format!("{base_name} (コピー {suffix})");
        if !exists(&candidate) {
            return Some(candidate);
        }
        suffix += 1;
    }
}
